docs/src/features/macros.md

```diff
@@ -0,0 +1,147 @@
+as the subject. A token tree is a sequence of tokens that are 
+enclosed in a pair of delimiters. Token trees are either `[...]`, `{...}` or `(...)`. It is then up to the macro to define various 
+rules for accepting the token tree:
+
+An example of using `min` and `max` macros:
+```rust
+main := () => {
+    min := @min {1 + 2, 3 * 4, 7 - 6 + 1 };
+    max := @max {1 + 2, 3 * 4, 7 - 6 + 1 };
+
+    if max - min == 0 {
+        println("min and max are equal")
+    } else {
+        println("min and max are not equal")
+    }
+}
+```
+
+Another example of macro with a token tree for HTML:
+```html
+welcome := () => {
+    @[xml(variant=html)] {
+        <html>
+            <head>
+                <title>My page</title>
+            </head>
+            <body>
+                <h1>Hello, world!</h1>
+            </body>
+        </html>
+    }
+}
+```
+
+
+## Defining a macro 🚧
+
+This section hasn't been defined yet, and is still a work in progress.
+
+
+## Macro Rules 🚧
+
+
+### Macro invocation locations  
+
+Both styles of macro invocations can appear in the following positions:
+- `Expr`
+- `Type`
+- `Pat`
+- `Param`
+- `Arg`
+- `TypeArg`
+- `PatArg`
+- `MatchCase`
+- `EnumVariant`
+
+Here is an example in code of all of the possible positions
+where a macro invocation can appear:
+
+```rust
+#![module_attributes]
+
+#dump_ast
+Foo := struct<#dump_ast T>(
+    #dump_ast x: T,
+    #dump_ast y: T,
+    #dump_ast z: T,
+);
+
+Bar := enum<T>(
+    #dump_ast A(T),
+    #dump_ast B(T),
+    #dump_ast C(T),
+);
+
+bing := (#param x: i32) -> #ty i32 => {
+    foo := Foo(#arg x = 5, #arg y = 6, #arg z = 7);
+
+    match x {
+        #dump_ast 0 => 0,
+        #dump_ast 1 => 1,
+        #dump_ast (#dump_ast _) => bing(x - 1) + bing(x - 2),
+    }
+}
+```
+
+
+### Syntax
+
+Formally, the macro syntax invocation can be written as follows:
+
+```bnf
+token_macro_invocation ::= "@" { macro_name | macro_args } token_tree;
+
+token_tree ::= "{" any "}" 
+             | "[" any "]" 
+             | "(" any ")";
+
+ast_macro_invocation ::= '#' (macro_name | macro_args ) macro_subject;
+
+module_macro_invocation ::= "#!" macro_args;
+
+macro_subject ::= expr 
+                  | type 
+                  | pat 
+                  | param 
+                  | arg 
+                  | type_arg
+                  | pat_arg
+                  | match_case 
+                  | enum_variant;  
+
+macro_args ::= "[" { ∅ | macro_invocation ("," macro_invocation)* ","? } "]";
+
+macro_invocation ::= macro_name ( "("  ∅ | expr ("," expr )* ","?  ")" )?;
+
+macro_name ::= ident;
```

Comment:
No namespacing?

Reply:
I was thinking about that, and I suppose ast path is better here.